src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,237 @@
+
+/-- The type of centroid homomorphisms from `α` to `α`. -/
+structure centroid_hom (α : Type*) [non_unital_non_assoc_semiring α] extends α →+ α :=
+(map_mul_left' (a b : α) : to_fun (a * b) = a * to_fun b)
+(map_mul_right' (a b : α) : to_fun (a * b) = to_fun a * b)
+
+/-- `centroid_hom_class F α` states that `F` is a type of centroid homomorphisms.
+
+You should extend this class when you extend `centroid_hom`. -/
+class centroid_hom_class (F : Type*) (α : out_param $ Type*) [non_unital_non_assoc_semiring α]
+  extends add_monoid_hom_class F α α :=
+(map_mul_left (f : F) (a b : α) : f (a * b) = a * f b)
+(map_mul_right (f : F) (a b : α) : f (a * b) = f a * b)
+
+export centroid_hom_class (map_mul_left map_mul_right)
+
+instance [non_unital_non_assoc_semiring α] [centroid_hom_class F α] :
+  has_coe_t F (centroid_hom α) :=
+⟨λ f, { to_fun := f, map_mul_left' := map_mul_left f, map_mul_right' := map_mul_right f,
+  ..(f : α →+ α) }⟩
+
+/-! ### Centroid homomorphisms -/
+
+namespace centroid_hom
+
+section non_unital_non_assoc_semiring
+
+variables [non_unital_non_assoc_semiring α]
+
+instance : centroid_hom_class (centroid_hom α) α :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by { obtain ⟨⟨_, _⟩, _⟩ := f, obtain ⟨⟨_, _⟩, _⟩ := g, congr' },
+  map_zero := λ f, f.map_zero',
+  map_add := λ f, f.map_add',
+  map_mul_left := λ f, f.map_mul_left',
+  map_mul_right := λ f, f.map_mul_right' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
+directly. -/
+instance : has_coe_to_fun (centroid_hom α) (λ _, α → α) := fun_like.has_coe_to_fun
+
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf] lemma to_fun_eq_coe {f : centroid_hom α} : f.to_fun = (f : α → α) := rfl
+
+@[ext] lemma ext {f g : centroid_hom α} (h : ∀ a, f a = g a) : f = g := fun_like.ext f g h
+
+@[simp] lemma coe_to_add_monoid_hom (f : centroid_hom α) : ⇑(f : α →+ α) = f := rfl
+@[simp] lemma to_add_monoid_hom_eq_coe (f : centroid_hom α) : f.to_add_monoid_hom = f :=
+fun_like.coe_injective rfl
+
+lemma coe_to_add_monoid_hom_injective : injective (coe : centroid_hom α → α →+ α) :=
+λ f g h, ext $ λ a, by { have := fun_like.congr_fun h a, exact this }
+
+/-- Turn a centroid homomorphism into an additive monoid endomorphism. -/
+def to_End (f : centroid_hom α) : add_monoid.End α := (f : α →+ α)
+
+lemma to_End_injective : injective (to_End : centroid_hom α → add_monoid.End α) :=
+coe_to_add_monoid_hom_injective
+
+/-- Copy of a `centroid_hom` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def copy (f : centroid_hom α) (f' : α → α) (h : f' = f) :
+  centroid_hom α :=
+{ to_fun := f',
+  map_mul_left' := λ a b, by simp_rw [h, map_mul_left],
+  map_mul_right' := λ a b, by simp_rw [h, map_mul_right],
+  ..f.to_add_monoid_hom.copy f' $ by exact h }
+
+variables (α)
+
+/-- `id` as a `centroid_hom`. -/
+protected def id : centroid_hom α := ⟨add_monoid_hom.id α, λ _ _, rfl, λ _ _, rfl⟩
```

Comment:
With the `old_structure_cmd` suggestion this becomes:
```suggestion
protected def id : centroid_hom α :=
{ map_mul_left' := λ _ _, rfl,
  map_mul_right' := λ _ _, rfl,
  .. add_monoid_hom.id α }
```

Reply:
In fact, you might want to apply all these structure definition suggestions even if we end up finding something better `old_structure_cmd`, since they are more robust against these kinds of changes.